### PR TITLE
[ZEPPELIN-819] Add 'Last updated by X at Y' message in paragraph footer

### PR DESCRIPTION
**Pull request by prasadwagle**
### What is this PR for?
In a large organization, it is helpful to know who updated a paragraph along with the time.

### What type of PR is it?
Improvement (Minor)

### Todos
* [x] - Understand why the user is not updated immediately after running the paragraph.

### What is the Jira issue?
[ZEPPELIN-819] (https://issues.apache.org/jira/browse/ZEPPELIN-819)

### How should this be tested?
1. Login to the zeppelin server
2. Create a note
3. Run a paragraph
4. Reload

Paragraph footer should display the 'Last updated by X at Y' message

### Screenshots (if appropriate)
NA

### Questions:
NA

### CODE REVIEW

File: zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js [hunk 1]
```diff
@@ -966,7 +966,7 @@ angular.module('zeppelinWebApp')
       }
       return '';
     }
-    var desc = 'Took ' + (timeMs/1000) + ' seconds';
+    var desc = 'Took ' + (timeMs/1000) + ' seconds. Last executed by ' + pdata.authenticationInfo.user + ' at ' + pdata.dateUpdated;
```
Comment (felixcheung) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js]: I think this should handle anonymous user case too?

Comment (AhyoungRyu) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js]: +1 for @felixcheung

Comment (prasadwagle) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js]: @felixcheung @AhyoungRyu I agree. I updated the PR. Can you please check if it looks ok? Thanks.

File: zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js [hunk 2]
```diff
@@ -966,7 +966,11 @@ angular.module('zeppelinWebApp')
       }
       return '';
     }
-    var desc = 'Took ' + (timeMs/1000) + ' seconds';
+    var user = 'anonymous';
+    if (pdata.authenticationInfo != null && pdata.authenticationInfo.user != null) {
+      user = pdata.authenticationInfo.user;
+    }
+    var desc = 'Took ' + (timeMs/1000) + ' seconds. Last updated by ' + user + ' at ' + pdata.dateUpdated;
```
Comment (corneadoug) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js]: I think that in the case of Zeppelin not using the auth, we should use the previous message (without anonymous)

**Comment by corneadoug** [comment]
I found a couple of issues:

1) The current message can show this:
`Took 0 seconds. Last updated by anonymous at undefined`
It would be better not to show `at undefined` in that case

2) `pdata.authenticationInfo.user` is not updated after running a paragraph

**Comment by Leemoonsoo** [comment]
Although last commit pass the CI, because https://github.com/apache/incubator-zeppelin/pull/854 is merged in master, merging this PR will fail with following error

```
[INFO] src/app/notebook/paragraph/paragraph.controller.js
[INFO]   line 972  col 36  Expected '!==' and instead saw '!='.
[INFO]   line 972  col 77  Expected '!==' and instead saw '!='.
[INFO]   line 975  col 44  Expected '===' and instead saw '=='.
[INFO] 
[INFO] ✖ 3 problems
```

@prasadwagle could you take care of them ?

**Comment by prasadwagle** [comment]
@Leemoonsoo I merged with latest master, fixed the jshint errors and tested with anonymous authentication. When I tried with basic authentication, I got:
GET http://localhost:8080/api/login;jsessionid=154khh1wznk3p2njo49c0dsea 405 (Method Not Allowed)
In the server log, I saw:
MacBook-Pro incubator-zeppelin-prasadwagle (ZEPPELIN-819) $ grep WARN logs/*
 WARN [2016-04-30 11:45:08,219] ({main} ServletHolder.java[getNameOfJspClass]:923) - Unable to make identifier for jsp rest trying rest instead
 WARN [2016-04-30 11:45:26,434] ({qtp2032251042-17} JAXRSUtils.java[findTargetMethod]:499) - No operation matching request path "/api/login;jsessionid=hmcl70a36eq4oii5owu4kvig" is found, Relative Path: /, HTTP Method: GET, ContentType: */*, Accept: text/html,application/xhtml+xml,image/webp,application/xml;q=0.9,*/*;q=0.8,. Please enable FINE/TRACE log level for more details.
 WARN [2016-04-30 11:45:26,435] ({qtp2032251042-17} WebApplicationExceptionMapper.java[toResponse]:73) - javax.ws.rs.ClientErrorException
I am debugging now. Am I doing something wrong?

**Comment by Leemoonsoo** [comment]
@prasadwagle I'm getting the same error. Trying to fix...

**Comment by Leemoonsoo** [comment]
@prasadwagle I have created fix for '405 Method Not Found'. Could you review https://github.com/apache/incubator-zeppelin/pull/870?

**Comment by prasadwagle** [comment]
@Leemoonsoo I made the change below from your PR 870 and the site works but I am still getting warnings in the log file.

diff --git a/zeppelin-server/src/main/java/org/apache/zeppelin/server/ZeppelinServer.java b/zeppelin-server/src/main/java/org/apache/zeppelin/server/ZeppelinServer.java
index f03fbeb..1730457 100644
--- a/zeppelin-server/src/main/java/org/apache/zeppelin/server/ZeppelinServer.java
+++ b/zeppelin-server/src/main/java/org/apache/zeppelin/server/ZeppelinServer.java
@@ -238,7 +238,7 @@ public class ZeppelinServer extends Application {
     webapp.setInitParameter("shiroConfigLocations",
         new File(conf.getShiroPath()).toURI().toString());
 
-    webapp.addFilter(org.apache.shiro.web.servlet.ShiroFilter.class, "/*",
+    webapp.addFilter(org.apache.shiro.web.servlet.ShiroFilter.class, "/api/*",
         EnumSet.allOf(DispatcherType.class));



 WARN [2016-05-02 13:35:59,068] ({qtp1725097945-37} JAXRSUtils.java[findTargetMethod]:499) - No operation matching request path "/api/login;jsessionid=5oi0fue75aua1s6u0r6jtsuja" is found, Relative Path: /, HTTP Method: GET, ContentType: */*, Accept:       application/json,text/plain,*/*,. Please enable FINE/TRACE log level for more details.
 WARN [2016-05-02 13:35:59,069] ({qtp1725097945-37} WebApplicationExceptionMapper.java[toResponse]:73) - javax.ws.rs.           ClientErrorException
        at org.apache.cxf.jaxrs.utils.JAXRSUtils.findTargetMethod(JAXRSUtils.java:503)
        at org.apache.cxf.jaxrs.interceptor.JAXRSInInterceptor.processRequest(JAXRSInInterceptor.java:227)
        at org.apache.cxf.jaxrs.interceptor.JAXRSInInterceptor.handleMessage(JAXRSInInterceptor.java:103)
        at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:272)
        at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
        at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:239)
        at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:248)
        at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:222)
        at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:153)
        at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:167)
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:286)
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doGet(AbstractHTTPServlet.java:211)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:575)
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:262)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)
        at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:61)
        at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
        at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
        at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
        at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
        at org.apache.shiro.web.servlet.AbstractShiroFilter.executeChain(AbstractShiroFilter.java:449)
        at org.apache.shiro.web.servlet.AbstractShiroFilter$1.call(AbstractShiroFilter.java:365)
        at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90)
        at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83)
        at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:383)
        at org.apache.shiro.web.servlet.AbstractShiroFilter.doFilterInternal(AbstractShiroFilter.java:362)
        at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
        at org.apache.zeppelin.server.CorsFilter.doFilter(CorsFilter.java:72)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:215)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
        at org.eclipse.jetty.server.Server.handle(Server.java:499)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
        at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
        at java.lang.Thread.run(Thread.java:745)

**Comment by Leemoonsoo** [comment]
The improvement looks good to me

**Comment by Leemoonsoo** [comment]
Merge if there're no more discussions

**Pull request closed without merging**